Add --rerunfailedsuites option

antechrestos opened the pull request:
Based on [this issue](https://github.com/robotframework/robotframework/issues/2117) (I do need this functionnality too), I started implementing it.

:warning: this do not support the `--critical/noncritical`as I **need some advice**:
- shall I filter from first output (where for example `--noncritical`was set) : then I need to update the model to get the information when output is processed
- shall I only filter from second execution parameters

Can you also give me some comment about my work, what I need to change/correct if any.

Thanks!

laurentbristiel commented on doc/userguide/src/ExecutingTestCases/ConfiguringExecution.rst:
```diff
@@ -165,6 +165,23 @@ running tests.
+Command line option :option:`rerunfailedsuites (-S)` can be used to select all failed
+suites from an earlier `output file`_ for re-execution. Like  :option:`--rerunfailed (-R)`, this option is useful when
+full test execution takes a lot of time. Note that all tests from a failed test suite will be re-executed, even
+passing ones. This option may be useful when you missed developing tests of a test suite as independent from each other.
```
"when you missed developing tests of a test suite as independent from each other"
=> bit unclear

antechrestos commented:
@laurentbristiel I replaced it with 
>>> This option is useful when the tests of a test suite depends on each other.

Do you have any idea about the advice I need for the `critical/noncritical`. It is not the way to do it, it is more what is expected.

Thanks!

antechrestos commented:
@pekkaklarck : I implemented it supposing the `--noncritical` tags had to be given at the second execution, and not guessed from the first output:
1. `pybot    ${SUITE DIR}`
2. `pybot --rerunfailedsuites ./output.xml --noncritical non_critical ${SUITE DIR}`
This will exclude tests bearing the *non_critical* tag from re execution (as described in `rerunfailedsuites.robot`).
 
Thinking this way, I found that the `critical` property of a `TestCase` was [a calculated property](https://github.com/robotframework/robotframework/blob/master/src/robot/result/model.py#L152) that bases its calculus upon parent (suite) `criticality` field. This later is set by the `Runner` ... I took a look at the `output.xml`and found no trace of the critical/noncritical tags. 

Consequently, I faced either overriding the `TestCase` property (which I found a little ugly), or building a `Criticability` at visit time. I then faced the last problem: when visiting, we are not sure that all settings property are set, specialy `critical` and `noncritical`. Hence, I implemented it as a trigger that get calculated at the end of `_BaseSettings._process_cli_opts`execution

Looking forward to reading you.

pekkaklarck commented on src/robot/conf/gatherfailed.py:
```diff
@@ -32,6 +32,21 @@ def visit_keyword(self, kw):
+    def __init__(self, critical_tags, non_critical_tags):
+        self.suites = set()
+        self._criticality = Criticality(critical_tags,
+                                        non_critical_tags)
```
This ought to work, but it might be cleaner to call `suite.set_crticality()` when `start_suite` is called for the first time. It would especially simplify `visit_test` implementation.

antechrestos replied:
@pekkaklarck yeah I first implemented this way but had a failure with my tests:
`Processing output failed: IOError: [Errno 2] No such file or directory: u'/tmp/robottests/Python-2.7-Linux/output/output.xml'`
looking closer at the documentation of showed me I missed the necessity to only set it on the root suite :face_with_thermometer:

pekkaklarck commented on src/robot/conf/settings.py:
```diff
@@ -92,6 +94,9 @@ def __setitem__(self, name, value):
             return gather_failed_tests(value)
+        if name == 'ReRunFailedSuites':
+            # will be calculated at the end
+            return lambda: gather_failed_suites(value, self.critical_tags, self.non_critical_tags)
```
Returning a lambda that is called in `_process_cli_opts()` is a bit ugly but works. Alternatively we could just let the normal logic set `ReRunFailedSuites` option and in `_process_cli_opts()` do something like this:

```python
if self['ReRunFailedSuites']:
    self['SuiteNames'] += gather_failued_suites(self['ReRunFailedSuites'])
```

For consistency reasons the logic with `ReRunFailed` should be changed to follow this logic too.

What do you think?

pekkaklarck replied:
Returning a lambda that is later called in `_process_cli_opts` is a bit ugly but works. Alternative option would be letting the normal option setting logic to set `ReRunFailedSuites` and then in `_process_cli_opts` doing something like this:

```python
if self['ReRunFailedSuites']:
    self['SuiteNames'] += gather_failed_suites(self['ReRunFailedSuites'], self.critical_tags, self.non_critical_tags)
```

For consistency reason `ReRunFailed` should be changed to follow the same logic. What do you think?

pekkaklarck commented:
Looks pretty good to me. Unfortunately I don't have time for a full review, but briefly looked and commented changes under `src`. I don't know when exactly I have time to look at this more closely (currently returning from a short holiday with too much tasks waiting for me...), but I'll anyway open the original issue and target it to RF 3.0.1. Thanks!

pekkaklarck commented:
Two more general comments before I need to close my laptop:

- Note(s) like "Added in Robot Framework 3.0.1" needed in docs.
- Add yourself to AUTHORS.txt if you want.

For more details see CONTRIBUTING.rst in the project root.

antechrestos commented:
@pekkaklarck I took into account your notes:
- I added a note in the *ConfiguringExecution.rst* file
 good luck for your return :smile:

pekkaklarck commented on doc/userguide/src/ExecutingTestCases/ConfiguringExecution.rst:
```diff
@@ -165,6 +165,25 @@ running tests.
+::
+
+  robot tests                                  # first execute all tests
+  robot --rerunfailedsuites output.xml tests    # then re-execute failing
```
Do we really need an example here when there's one with `--rerunfailed` and these features work exactly same way?

pekkaklarck commented on atest/robot/cli/runner/rerunfailedsuites.robot:
```diff
@@ -0,0 +1,44 @@
+
+*** Variables ***
+${ORIG DIR}           ${DATADIR}/cli/runfailedsuites
+${RERUN SUITE DIR}          %{TEMPDIR}/rerunsuites-dir
```
Strange amount of spaces here. Or are there tabs?

pekkaklarck commented on doc/userguide/src/Appendices/CommandLineOptions.rst:
```diff
@@ -22,6 +22,8 @@ Command line options for test execution
   -R, --rerunfailed <file>  `Selects failed tests`_ from an earlier `output file`_ to be re-executed.
   --runfailed <file>      Deprecated since Robot Framework 2.8.4.
                           Use :option:`--rerunfailed` instead.
+  -S, --rerunfailedsuites <file>  `Selects failed test suite`_ from an earlier `output file`_ to be re-executed.
```
This link doesn't match the target. I'll fix it myself.

pekkaklarck commented:
Reviewing the this PR was a pleasure. I can be pretty nitpicky in my comments, but had absolutely nothing to say about core or tests in the final version. The was one broken link in docs but that was easy to fix. Thanks @antechrestos!

antechrestos commented:
@pekkaklarck no problem. In fact I like this way because it allows to share knowledge.  Coding this feature brought me the opportunity to look deeper into the beast :smirk:. 
Being nitpicky as you say is useful as it maintains a good quality of code and I respect that.
